dspace-api/src/main/java/org/dspace/submit/consumer/SubmissionConfigConsumer.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.submit.consumer;
+
+import org.apache.logging.log4j.Logger;
+import org.dspace.content.Collection;
+import org.dspace.content.DSpaceObject;
+import org.dspace.core.Constants;
+import org.dspace.core.Context;
+import org.dspace.discovery.IndexingService;
+import org.dspace.discovery.indexobject.IndexableCollection;
+import org.dspace.event.Consumer;
+import org.dspace.event.Event;
+import org.dspace.services.factory.DSpaceServicesFactory;
+import org.dspace.submit.factory.SubmissionServiceFactory;
+
+/**
+ * Consumer implementation to be used for Item Submission Configuration
+ *
+ * @author paulo.graca at fccn.pt
+ */
+public class SubmissionConfigConsumer implements Consumer {
+    /**
+     * log4j logger
+     */
+    private static Logger log = org.apache.logging.log4j.LogManager.getLogger(SubmissionConfigConsumer.class);
+
+    IndexingService indexer = DSpaceServicesFactory.getInstance().getServiceManager()
+            .getServiceByName(IndexingService.class.getName(),
+                              IndexingService.class);
+
+    @Override
+    public void initialize() throws Exception {
+        // No-op
+    }
+
+    @Override
+    public void consume(Context ctx, Event event) throws Exception {
+        int st = event.getSubjectType();
+        int et = event.getEventType();
+
+
+        if ( st == Constants.COLLECTION ) {
+            switch (et) {
+                case Event.MODIFY_METADATA:
+                    // Submission configuration it's based on solr
+                    // for collection's entity type but, at this point
+                    // that info isn't indexed yet, we need to force it
+                    DSpaceObject subject = event.getSubject(ctx);
+                    Collection collectionFromDSOSubject = (Collection) subject;
+                    indexer.indexContent(ctx, new IndexableCollection (collectionFromDSOSubject), true, false, false);
+                    indexer.commit();
```

Comment:
I'm a little confused by the need to force reindexing of the Collection into Solr.  I had always thought that the Event Consumers listed in `event.dispatcher.default.consumers` were run *in order*.  So, as long as the `discovery` consumer is listed _before_ the `submissionconfig` consumer, then shouldn't this Collection already be indexed?

Maybe I'm missing something though, or maybe these consumers are no longer run in order, or there's some sort of "race condition" here?

Reply:
I think it's related with the way IndexEventConsumer is implemented. The Solr `commit` will only occur in the `end` method, not in the `consume`:
https://github.com/paulo-graca/DSpace/blob/enhancement/itemsubmission-config-reload-consumer/dspace-api/src/main/java/org/dspace/discovery/IndexEventConsumer.java#L202-L234